src/main/scala/viper/gobra/frontend/Parser.scala

```diff
@@ -421,11 +421,30 @@ object Parser {
           PFunctionDecl(name, sig._1, sig._2, spec, body)
       }
 
-    lazy val functionSpec: Parser[PFunctionSpec] =
-      ("requires" ~> expression <~ eos).* ~ ("ensures" ~> expression <~ eos).* ~ "pure".? ^^ {
-        case pres ~ posts ~ isPure => PFunctionSpec(pres, posts, isPure.nonEmpty)
+
+
+    lazy val functionSpec: Parser[PFunctionSpec] = {
+
+      sealed trait FunctionSpecClause
+      case class RequiresClause(exp: PExpression) extends FunctionSpecClause
+      case class EnsuresClause(exp: PExpression) extends FunctionSpecClause
+      case object PureClause extends FunctionSpecClause
+
+      lazy val functSpecClause: Parser[FunctionSpecClause] = {
+        "requires" ~> expression <~ eos ^^ RequiresClause |
+        "ensures" ~> expression <~ eos ^^ EnsuresClause |
+        "pure" <~ eos ^^^ PureClause
       }
 
+      functSpecClause.* ~ "pure".? ^^ {
+        case clauses ~ pure =>
+          val pres = clauses.collect{ case x: RequiresClause => x.exp }
+          val posts = clauses.collect{ case x: EnsuresClause => x.exp }
+          val isPure = pure.nonEmpty || clauses.contains(PureClause)
+          PFunctionSpec(pres, posts, isPure)
```

Comment:
shouldn't you attach positional information to `PFunctionSpec` here?

Reply:
that is done by the parser framework